# intro

## about dataset

The dataset was created in a project that aims to contribute to the reduction of academic dropout and failure in higher education, by using machine learning techniques to identify students at risk at an early stage of their academic path, so that strategies to support them can be put into place. 

## where to download

<p>you can dwonlad dataset from <a href="https://archive.ics.uci.edu/dataset/697/predict+students+dropout+and+academic+success">here<a/>.</p>

# understand dataset

## import libraries

In [19]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

## read dataset

In [20]:
df = pd.read_csv(filepath_or_buffer="./data.csv", sep=";")
df.head()

Marital status  Application mode  Application order  Course  \
0               1                17                  5     171   
1               1                15                  1    9254   
2               1                 1                  5    9070   
3               1                17                  2    9773   
4               2                39                  1    8014   

   Daytime/evening attendance\t  Previous qualification  \
0                             1                       1   
1                             1                       1   
2                             1                       1   
3                             1                       1   
4                             0                       1   

   Previous qualification (grade)  Nacionality  Mother's qualification  \
0                           122.0            1                      19   
1                           160.0            1                       1   
2                           122.0            1                      37   
3                           122.0            1                      38   
4                           100.0            1                      37   

   Father's qualification  ...  Curricular units 2nd sem (credited)  \
0                      12  ...                                    0   
1                       3  ...                                    0   
2                      37  ...                                    0   
3                      37  ...                                    0   
4                      38  ...                                    0   

   Curricular units 2nd sem (enrolled)  \
0                                    0   
1                                    6   
2                                    6   
3                                    6   
4                                    6   

   Curricular units 2nd sem (evaluations)  \
0                                       0   
1                                       6   
2                                       0   
3                                      10   
4                                       6   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                    0                          0.000000   
1                                    6                         13.666667   
2                                    0                          0.000000   
3                                    5                         12.400000   
4                                    6                         13.000000   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
0                                               0               10.8   
1                                               0               13.9   
2                                               0               10.8   
3                                               0                9.4   
4                                               0               13.9   

   Inflation rate   GDP    Target  
0             1.4  1.74   Dropout  
1            -0.3  0.79  Graduate  
2             1.4  1.74   Dropout  
3            -0.8 -3.12  Graduate  
4            -0.3  0.79  Graduate  

[5 rows x 37 columns]

## get some info about featuers

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   int64  
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4424 non-null   int64  
 4   Daytime/evening attendance	                     4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

## get dataset features name

In [22]:
df.columns

Index(['Marital status', 'Application mode', 'Application order', 'Course',
       'Daytime/evening attendance\t', 'Previous qualification',
       'Previous qualification (grade)', 'Nacionality',
       'Mother's qualification', 'Father's qualification',
       'Mother's occupation', 'Father's occupation', 'Admission grade',
       'Displaced', 'Educational special needs', 'Debtor',
       'Tuition fees up to date', 'Gender', 'Scholarship holder',
       'Age at enrollment', 'International',
       'Curricular units 1st sem (credited)',
       'Curricular units 1st sem (enrolled)',
       'Curricular units 1st sem (evaluations)',
       'Curricular units 1st sem (approved)',
       'Curricular units 1st sem (grade)',
       'Curricular units 1st sem (without evaluations)',
       'Curricular units 2nd sem (credited)',
       'Curricular units 2nd sem (enrolled)',
       'Curricular units 2nd sem (evaluations)',
       'Curricular units 2nd sem (approved)',
       'Curricular units 2nd

## check for null and nan

In [23]:
print('check nan:')
print(df.isna().sum())
print('\n\ncheck null:')
print(df.isnull().sum())

check nan:
Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            0
Daytime/evening attendance\t                      0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            0
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   0
Displaced                                         0
Educational special needs                         0
Debtor                                            0
Tuition fees up to date                           0
Gender                                            0
Scholarship holder                                0
A

## final analysis

- **about dataset**:
    1. we've got 4424 rows with 37 columns
    2. only target column is categorical
    3. there is no null or NaN field in dataset

- **about features**:

    - **demographic features**:
        1. **marital_status**: indicates the student's marital status, which may impact time availability and responsibilities.
        2. **gender**: student's gender, which may reveal patterns in educational outcomes across different genders.
        3. **age_at_enrollment**: age when the student enrolled, potentially affecting maturity, preparation, and commitment levels.
        4. **nationality**: student's country of origin, which may relate to cultural background and adaptation challenges.
        5. **is_international**: binary indicator of whether the student is international, related to potential language barriers and adaptation issues.
        6. **is_displaced**: indicates if the student is studying away from their home region, which may affect support systems and adaptation.
        7. **has_special_needs**: identifies students with educational special needs requiring additional support.
        
    - **family background**:
        8. **mother_qual**: mother's educational qualification level, an important socioeconomic indicator.
        9. **father_qual**: father's educational qualification level, another key socioeconomic indicator.
        10. **mother_occupation**: mother's professional occupation, reflecting family socioeconomic status.
        11. **father_occupation**: father's professional occupation, also reflecting family socioeconomic status.
        
    - **application and enrollment**:
        12. **app_mode**: how the student applied to the program (e.g., standard, special admission).
        13. **app_order**: the order of preference the student gave this course in their application, indicating motivation level.
        14. **course**: the specific program the student is enrolled in.
        15. **attendance_type**: whether the student attends daytime or evening classes, reflecting potential work commitments.
        16. **prev_qualification**: the type of qualification the student had before entering higher education.
        17. **prev_qual_grade**: the grade achieved in their previous qualification, indicating prior academic performance.
        
    - **financial factors**:
        18. **is_debtor**: whether the student has outstanding debt to the institution.
        19. **fees_up_to_date**: whether the student's tuition payments are current.
        20. **has_scholarship**: whether the student receives financial aid, indicating financial need and/or academic merit.
        
    - **academic performance (first semester)**:
        21. **sem1_credits**: number of course units credited/recognized from previous studies.
        22. **sem1_enrolled**: number of curricular units the student enrolled in.
        23. **sem1_evaluations**: number of evaluations the student participated in.
        24. **sem1_approved**: number of course units the student successfully passed.
        25. **sem1_grade**: average grade across all units in first semester.
        26. **sem1_without_evals**: number of units where the student didn't participate in evaluations.
        
    - **academic performance (second semester)**:
        27. **sem2_credits**: number of course units credited from previous studies.
        28. **sem2_enrolled**: number of curricular units the student enrolled in.
        29. **sem2_evaluations**: number of evaluations the student participated in.
        30. **sem2_approved**: number of course units the student successfully passed.
        31. **sem2_grade**: average grade across all units in second semester.
        32. **sem2_without_evals**: number of units where the student didn't participate in evaluations.
        
    - **economic context**:
        33. **unemployment_rate**: national unemployment rate when the student was enrolled, reflecting job market conditions.
        34. **inflation_rate**: inflation rate when the student was enrolled, indicating economic stability.
        35. **gdp**: gross domestic product when the student was enrolled, reflecting overall economic health.
        
    - **target variable**:
        36. **target**: the outcome variable indicating whether the student dropped out, graduated, or is still enrolled.

## rename columns

In [ ]:
column_mapping = {
    'Marital status': 'marital_status',
    'Application mode': 'app_mode',
    'Application order': 'app_order',
    'Course': 'course',
    'Daytime/evening attendance\t': 'attendance_type',
    'Previous qualification': 'prev_qualification',
    'Previous qualification (grade)': 'prev_qual_grade',
    'Nacionality': 'nationality',
    "Mother's qualification": 'mother_qual',
    "Father's qualification": 'father_qual',
    "Mother's occupation": 'mother_occupation',
    "Father's occupation": 'father_occupation',
    'Admission grade': 'admission_grade',
    'Displaced': 'is_displaced',
    'Educational special needs': 'has_special_needs',
    'Debtor': 'is_debtor',
    'Tuition fees up to date': 'fees_up_to_date',
    'Gender': 'gender',
    'Scholarship holder': 'has_scholarship',
    'Age at enrollment': 'age_at_enrollment',
    'International': 'is_international',
    'Curricular units 1st sem (credited)': 'sem1_credits',
    'Curricular units 1st sem (enrolled)': 'sem1_enrolled',
    'Curricular units 1st sem (evaluations)': 'sem1_evaluations',
    'Curricular units 1st sem (approved)': 'sem1_approved',
    'Curricular units 1st sem (grade)': 'sem1_grade',
    'Curricular units 1st sem (without evaluations)': 'sem1_without_evals',
    'Curricular units 2nd sem (credited)': 'sem2_credits',
    'Curricular units 2nd sem (enrolled)': 'sem2_enrolled',
    'Curricular units 2nd sem (evaluations)': 'sem2_evaluations',
    'Curricular units 2nd sem (approved)': 'sem2_approved',
    'Curricular units 2nd sem (grade)': 'sem2_grade',
    'Curricular units 2nd sem (without evaluations)': 'sem2_without_evals',
    'Unemployment rate': 'unemployment_rate',
    'Inflation rate': 'inflation_rate',
    'GDP': 'gdp',
    'Target': 'target'
}

df = df.rename(columns=column_mapping)
df.head()

## save correlation result as csv for statistic analysing

In [25]:
corr_result = df.corr(numeric_only=True)
corr_result.to_csv('correlation_matrix.csv', float_format='%.2f')